# California Housing Price Prediction

**Project goal:** Building and comparing two linear regression models for predicting the median house value (`median_house_value`) on the California Housing 1990 data.

**Technology stack:** PySpark 3.x, MLlib, DataFrame API

**Target variable:** `median_house_value`

**Evaluation metrics:** RMSE, MAE, R²

**Models:**
- **Model 1:** All features (numerical + OHE-encoded `ocean_proximity`)
- **Model 2:** Numerical features only

## Imports

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, isnan, when, count, mean
from pyspark.sql.types import StringType

from pyspark.ml import Pipeline
from pyspark.ml.feature import StringIndexer, OneHotEncoder, VectorAssembler, StandardScaler, Imputer
from pyspark.ml.regression import LinearRegression
from pyspark.ml.evaluation import RegressionEvaluator

import pandas as pd

/home/yka/anaconda3/envs/practicum/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 2.0.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


## Global Constants

In [2]:
RANDOM_SEED = 42
DATA_PATH = "source/housing.csv"

## Spark Session Initialization

In [3]:
spark = (
    SparkSession
    .builder
    .appName("California Housing - Linear Regression")
    .master("local[3]")
    .config("spark.driver.memory", "6g")
    .config("spark.executor.memory", "2g")
    .config("spark.sql.shuffle.partitions", "8")
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    .getOrCreate()
)

print(f"Spark version: {spark.version}")

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
26/01/17 11:53:13 WARN Utils: Your hostname, EvanderPC, resolves to a loopback address: 127.0.1.1; using 10.255.255.254 instead (on interface lo)
26/01/17 11:53:13 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
26/01/17 11:53:23 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Spark version: 4.0.0


### Result

SparkSession was successfully initialized with a local configuration (3 threads, 6GB driver memory).

## Loading and Initial Data Inspection

In [4]:
try:
    df = spark.read.csv(DATA_PATH, header = True, inferSchema = True)
except:
    df = spark.read.csv("/datasets/housing.csv", header = True, inferSchema = True)

In [5]:
df.printSchema()

root
 |-- longitude: double (nullable = true)
 |-- latitude: double (nullable = true)
 |-- housing_median_age: double (nullable = true)
 |-- total_rooms: double (nullable = true)
 |-- total_bedrooms: double (nullable = true)
 |-- population: double (nullable = true)
 |-- households: double (nullable = true)
 |-- median_income: double (nullable = true)
 |-- median_house_value: double (nullable = true)
 |-- ocean_proximity: string (nullable = true)



In [6]:
df.show(10)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
|  -122.23|   37.88|              41.0|      880.0|         129.0|     322.0|     126.0|       8.3252|          452600.0|       NEAR BAY|
|  -122.22|   37.86|              21.0|     7099.0|        1106.0|    2401.0|    1138.0|       8.3014|          358500.0|       NEAR BAY|
|  -122.24|   37.85|              52.0|     1467.0|         190.0|     496.0|     177.0|       7.2574|          352100.0|       NEAR BAY|
|  -122.25|   37.85|              52.0|     1274.0|         235.0|     558.0|     219.0|       5.6431|          341300.0|       NEAR BAY|
|  -122.25|   37.85|              

In [7]:
display(df.describe().toPandas())

26/01/17 11:53:28 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


,summary,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,count,20640,20640,20640,20640,20433,20640,20640,20640,20640,20640
1,mean,-119.56970445736148,35.6318614341087,28.639486434108527,2635.7630813953488,537.8705525375618,1425.4767441860465,499.5396802325581,3.8706710029070246,206855.81690891474,None
2,stddev,2.003531723502584,2.135952397457101,12.58555761211163,2181.6152515827944,421.38507007403115,1132.46212176534,382.3297528316098,1.899821717945263,115395.61587441359,None
3,min,-124.35,32.54,1.0,2.0,1.0,3.0,1.0,0.4999,14999.0,<1H OCEAN
4,max,-114.31,41.95,52.0,39320.0,6445.0,35682.0,6082.0,15.0001,500001.0,NEAR OCEAN


In [8]:
print(f"Количество строк: {df.count()}")
print(f"Количество колонок: {len(df.columns)}")

Количество строк: 20640
Количество колонок: 10


### Result

The dataset has been loaded. The schema matches the expected one: 8 numerical features (double), 1 categorical (string), 1 target variable (double).

## Missing Values Analysis

In [9]:
numeric_cols_check = [f.name for f in df.schema.fields if not isinstance(f.dataType, StringType)]
string_cols_check = [f.name for f in df.schema.fields if isinstance(f.dataType, StringType)]

null_exprs = []
for c in numeric_cols_check:
    null_exprs.append(count(when(col(c).isNull() | isnan(c), c)).alias(c))
for c in string_cols_check:
    null_exprs.append(count(when(col(c).isNull(), c)).alias(c))

null_counts = df.select(null_exprs)
display(null_counts.toPandas())

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,0,0,0,0,207,0,0,0,0,0


### Result

Missing values were found in the `total_bedrooms` column (207 values). Filling the missing values will be performed with `Imputer` inside the Pipeline, so that the median is computed only on the training set.

## Defining the Feature Lists

In [10]:
categorical_cols = ["ocean_proximity"]

numerical_cols = [
    "longitude",
    "latitude",
    "housing_median_age",
    "total_rooms",
    "total_bedrooms_imputed",
    "population",
    "households",
    "median_income"
]

target = "median_house_value"

print(f"Категориальные признаки: {categorical_cols}")
print(f"Числовые признаки: {numerical_cols}")
print(f"Целевая переменная: {target}")

Категориальные признаки: ['ocean_proximity']
Числовые признаки: ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms_imputed', 'population', 'households', 'median_income']
Целевая переменная: median_house_value


### Result

The feature lists have been defined. The target variable is excluded from the feature list. Please note: `total_bedrooms_imputed` is the column name after applying `Imputer`.

## Splitting into Training and Test Sets

In [11]:
train_data, test_data = df.randomSplit([0.8, 0.2], seed = RANDOM_SEED)

In [12]:
train_count = train_data.count()
test_count = test_data.count()
total_count = train_count + test_count

print(f"Обучающая выборка: {train_count} ({train_count / total_count * 100:.1f}%)")
print(f"Тестовая выборка: {test_count} ({test_count / total_count * 100:.1f}%)")

Обучающая выборка: 16560 (80.2%)
Тестовая выборка: 4080 (19.8%)


### Result

The data has been split into training (80%) and test (20%) sets with a fixed `RANDOM_SEED = 42` for reproducibility. **Important:** the split was performed BEFORE any transformations, in order to avoid data leakage.

## Viewing the Unique Values of the Categorical Feature

In [13]:
df.select("ocean_proximity").distinct().show()

+---------------+
|ocean_proximity|
+---------------+
|     NEAR OCEAN|
|         ISLAND|
|      <1H OCEAN|
|         INLAND|
|       NEAR BAY|
+---------------+



## Creating the Pipeline for Model 1 (all features)

In [14]:
imputer = Imputer(
    inputCols = ["total_bedrooms"],
    outputCols = ["total_bedrooms_imputed"],
    strategy = "median"
)

indexer = StringIndexer(
    inputCol = "ocean_proximity",
    outputCol = "ocean_proximity_idx",
    handleInvalid = "keep"
)

encoder = OneHotEncoder(
    inputCol = "ocean_proximity_idx",
    outputCol = "ocean_proximity_ohe",
    dropLast = True
)

numerical_assembler = VectorAssembler(
    inputCols = numerical_cols,
    outputCol = "numerical_features",
    handleInvalid = "keep"
)

scaler = StandardScaler(
    inputCol = "numerical_features",
    outputCol = "numerical_features_scaled",
    withMean = True,
    withStd = True
)

all_features_assembler = VectorAssembler(
    inputCols = ["numerical_features_scaled", "ocean_proximity_ohe"],
    outputCol = "features_all",
    handleInvalid = "keep"
)

lr_all = LinearRegression(
    featuresCol = "features_all",
    labelCol = "median_house_value",
    predictionCol = "prediction_all",
    regParam = 0.01
)

pipeline_all = Pipeline(stages = [
    imputer,
    indexer,
    encoder,
    numerical_assembler,
    scaler,
    all_features_assembler,
    lr_all
])

print("Pipeline для Модели 1 создан.")
print(f"Этапы: {[type(stage).__name__ for stage in pipeline_all.getStages()]}")

Pipeline для Модели 1 создан.
Этапы: ['Imputer', 'StringIndexer', 'OneHotEncoder', 'VectorAssembler', 'StandardScaler', 'VectorAssembler', 'LinearRegression']


### Result

The Pipeline for Model 1 has been created with the stages:
1. **Imputer** — filling missing values with the median (computed on train)
2. **StringIndexer** — converting `ocean_proximity` into a numerical index
3. **OneHotEncoder** — OHE encoding of the categorical feature
4. **VectorAssembler** — assembling the numerical features into a vector
5. **StandardScaler** — standardization of the numerical features
6. **VectorAssembler** — combining the scaled numerical and OHE features
7. **LinearRegression** — linear regression model with regularization (regParam=0.01)

## Training Model 1

In [15]:
model_all = pipeline_all.fit(train_data)

In [16]:
lr_model_all = model_all.stages[-1]
print(f"Коэффициенты модели 1: {lr_model_all.coefficients}")
print(f"Intercept модели 1: {lr_model_all.intercept}")

Коэффициенты модели 1: [-53605.17300751415,-54869.991032304526,13275.040289251223,-7513.85742519107,28868.39047817371,-49650.745963730114,33427.97475083084,72436.53419731578,12507.472427689148,-26619.96910052803,17479.26153120307,7379.132430573913,162240.91365462114]
Intercept модели 1: 205552.11033569011


In [17]:
predictions_all = model_all.transform(test_data)
predictions_all.select("median_house_value", "prediction_all").show(10)

+------------------+------------------+
|median_house_value|    prediction_all|
+------------------+------------------+
|          103600.0|150535.19525634902|
|          106700.0| 217627.6614955553|
|           73200.0|125170.96373121833|
|           90100.0| 195021.5587143444|
|           67000.0|152292.80485485384|
|           86400.0| 186078.3090950679|
|           70500.0|163940.37848860433|
|           85100.0|180011.97785721224|
|           80500.0| 181582.5950584301|
|           96000.0|170317.42311325346|
+------------------+------------------+
only showing top 10 rows


### Result

Model 1 (all features: numerical + OHE-encoded `ocean_proximity`) has been trained through the Pipeline. All transformers were fitted on the training data. Predictions were obtained on the test set.

## Creating the Pipeline for Model 2 (numerical features only)

In [18]:
imputer_2 = Imputer(
    inputCols = ["total_bedrooms"],
    outputCols = ["total_bedrooms_imputed"],
    strategy = "median"
)

numerical_assembler_2 = VectorAssembler(
    inputCols = numerical_cols,
    outputCol = "numerical_features",
    handleInvalid = "keep"
)

scaler_2 = StandardScaler(
    inputCol = "numerical_features",
    outputCol = "features_numerical_only",
    withMean = True,
    withStd = True
)

lr_num = LinearRegression(
    featuresCol = "features_numerical_only",
    labelCol = "median_house_value",
    predictionCol = "prediction_num",
    regParam = 0.01
)

pipeline_num = Pipeline(stages = [
    imputer_2,
    numerical_assembler_2,
    scaler_2,
    lr_num
])

print("Pipeline для Модели 2 создан.")
print(f"Этапы: {[type(stage).__name__ for stage in pipeline_num.getStages()]}")

Pipeline для Модели 2 создан.
Этапы: ['Imputer', 'VectorAssembler', 'StandardScaler', 'LinearRegression']


### Result

The Pipeline for Model 2 has been created with the stages:
1. **Imputer** — filling missing values with the median
2. **VectorAssembler** — assembling the numerical features into a vector
3. **StandardScaler** — standardization (the output is used directly as features)
4. **LinearRegression** — linear regression model with regularization (regParam=0.01)

## Training Model 2

In [19]:
model_num = pipeline_num.fit(train_data)

In [20]:
lr_model_num = model_num.stages[-1]
print(f"Коэффициенты модели 2: {lr_model_num.coefficients}")
print(f"Intercept модели 2: {lr_model_num.intercept}")

Коэффициенты модели 2: [-85346.34258290856,-91314.18863638745,14233.172317804969,-11463.122268839223,32852.98027905642,-50250.3952904399,33862.920291080096,74230.51348567121]
Intercept модели 2: 205596.3980676288


In [21]:
predictions_num = model_num.transform(test_data)
predictions_num.select("median_house_value", "prediction_num").show(10)

+------------------+------------------+
|median_house_value|    prediction_num|
+------------------+------------------+
|          103600.0|100751.62819965834|
|          106700.0| 190789.9911644132|
|           73200.0|  74735.1749868445|
|           90100.0|162343.73090071618|
|           67000.0|119470.37450376325|
|           86400.0|155923.36849880053|
|           70500.0| 131197.3884082705|
|           85100.0|150449.15803565722|
|           80500.0|150168.96489164347|
|           96000.0|133760.90806567276|
+------------------+------------------+
only showing top 10 rows


### Result

Model 2 (numerical features only) has been trained through the Pipeline. Predictions were obtained on the test set.

## Evaluating Model Quality

In [22]:
evaluator_rmse_all = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_all",
    metricName = "rmse"
)
evaluator_mae_all = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_all",
    metricName = "mae"
)
evaluator_r2_all = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_all",
    metricName = "r2"
)

rmse_all = evaluator_rmse_all.evaluate(predictions_all)
mae_all = evaluator_mae_all.evaluate(predictions_all)
r2_all = evaluator_r2_all.evaluate(predictions_all)

print("Модель 1 (все признаки):")
print(f"  RMSE: {rmse_all:,.2f}")
print(f"  MAE:  {mae_all:,.2f}")
print(f"  R²:   {r2_all:.4f}")

Модель 1 (все признаки):
  RMSE: 70,786.68
  MAE:  50,863.76
  R²:   0.6378


In [23]:
evaluator_rmse_num = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_num",
    metricName = "rmse"
)
evaluator_mae_num = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_num",
    metricName = "mae"
)
evaluator_r2_num = RegressionEvaluator(
    labelCol = "median_house_value",
    predictionCol = "prediction_num",
    metricName = "r2"
)

rmse_num = evaluator_rmse_num.evaluate(predictions_num)
mae_num = evaluator_mae_num.evaluate(predictions_num)
r2_num = evaluator_r2_num.evaluate(predictions_num)

print("Модель 2 (только числовые):")
print(f"  RMSE: {rmse_num:,.2f}")
print(f"  MAE:  {mae_num:,.2f}")
print(f"  R²:   {r2_num:.4f}")

Модель 2 (только числовые):
  RMSE: 71,791.59
  MAE:  51,804.75
  R²:   0.6275


In [24]:
results_df = pd.DataFrame({
    "Модель": ["Модель 1 (все признаки)", "Модель 2 (числовые)"],
    "RMSE": [rmse_all, rmse_num],
    "MAE": [mae_all, mae_num],
    "R²": [r2_all, r2_num]
})

display(results_df)

,Модель,RMSE,MAE,R²
0,Модель 1 (все признаки),70786.679706,50863.756294,0.637843
1,Модель 2 (числовые),71791.593718,51804.750819,0.627487


### Result

The RMSE, MAE and R² metrics have been calculated for both models and presented in a comparison table.

## Comparative Analysis

In [25]:
rmse_diff = rmse_num - rmse_all
rmse_improvement = (rmse_diff / rmse_num) * 100

r2_diff = r2_all - r2_num

print("Сравнительный анализ:")
print(f"  Разница RMSE: {rmse_diff:,.2f} (улучшение {rmse_improvement:.2f}%)")
print(f"  Разница R²: {r2_diff:.4f}")

Сравнительный анализ:
  Разница RMSE: 1,004.91 (улучшение 1.40%)
  Разница R²: 0.0104


## Overall Conclusion

### Work Done

Within this project, the task of predicting the median house value in California (`median_house_value`) was solved on the basis of the California Housing 1990 dataset using PySpark MLlib.

**Main stages:**
1. **Loading and exploring the data:** A dataset of 20,640 records with 9 features (8 numerical + 1 categorical `ocean_proximity`) was analyzed.

2. **Handling missing values:** 207 missing values were found in the `total_bedrooms` column. To fill them, `Imputer` inside the Pipeline with the median strategy was used, which guarantees that the median is computed only on the training data.

3. **Data splitting:** The dataset was split into training (80%) and test (20%) sets BEFORE applying any transformations, in order to exclude data leakage.

4. **Building the Pipeline:** For each model a separate Pipeline was created, including all preprocessing stages (Imputer, StringIndexer, OneHotEncoder, VectorAssembler, StandardScaler) and training (LinearRegression with L2 regularization regParam=0.01).

5. **Training and evaluating the models:**
   - **Model 1** (all features): uses the numerical and the OHE-encoded categorical feature
   - **Model 2** (numerical only): uses only the numerical features

### Results

| Metric | Model 1 (all features) | Model 2 (numerical) |
|---------|------------------------|--------------------|
| RMSE    | ~70,000-71,000         | ~71,000-72,000     |
| MAE     | ~50,000-51,000         | ~51,000-52,000     |
| R²      | ~0.63-0.64             | ~0.62-0.63         |

**Impact of the categorical feature:**
Adding the OHE-encoded feature `ocean_proximity` (proximity to the ocean) improves the model quality by approximately 1-1.5% in RMSE and ~1% in R². This is logical, since the location relative to the ocean is a significant factor of real estate value in California.

### Recommendations for Improvement

1. **Feature Engineering:**
   - Creating derived features: `rooms_per_household = total_rooms / households`, `bedrooms_ratio = total_bedrooms / total_rooms`
   - Binarization or polynomial transformation of features

2. **Hyperparameter search:**
   - Using CrossValidator to optimize `regParam` and `elasticNetParam`
   - Experiments with different types of regularization (L1, L2, ElasticNet)

3. **Alternative models:**
   - Gradient Boosted Trees (GBTRegressor)
   - Random Forest (RandomForestRegressor)
   - These models capture nonlinear dependencies better

### Limitations of Linear Regression

An R² of about 0.63 indicates that the linear model explains only ~63% of the variance of the target variable. This may be related to:
- Nonlinear dependencies between the features and the target variable
- Spatial autocorrelation (neighboring districts have similar prices)
- The target variable being capped from above (values capped at 500,001)

In [26]:
spark.stop()